llvm/lib/Target/AIE/AIECombinerHelper.cpp

```diff
@@ -1119,3 +1121,190 @@ void llvm::applyPadVector(MachineInstr &MI, MachineRegisterInfo &MRI,
+      TII.getVExtractOpInfo(MI);
+
+  if (!ExtractOp)
+    return false;
+
+  auto Cst = getIConstantVRegValWithLookThrough(
+      MI.getOperand(ExtractOp->SubVectorIndex).getReg(), MRI);
+
+  if (!Cst)
+    return false;
+
+  const unsigned ExtractSize =
+      MRI.getType(MI.getOperand(0).getReg()).getSizeInBits();
+
+  MachineInstr &SrcMI = *MRI.getVRegDef(MI.getOperand(ExtractOp->Src).getReg());
+
+  Register SrcReg;
+  unsigned ConcatSize = 0;
+  const unsigned Index = Cst->Value.getZExtValue();
+
+  if (const auto ConcatOp = TII.getVConcatOpInfo(SrcMI)) {
+    SrcReg = SrcMI.getOperand(Index + ConcatOp->FirstOperand).getReg();
+    ConcatSize = MRI.getType(SrcReg).getSizeInBits();
+  }
+
+  MatchInfo = SrcReg;
+  return ConcatSize == ExtractSize;
+}
+
+void llvm::applyExtractConcat(MachineInstr &MI, MachineRegisterInfo &MRI,
+                              MachineIRBuilder &B, Register &MatchInfo) {
+  B.setInstrAndDebugLoc(MI);
+  Register DstReg = MI.getOperand(0).getReg();
+  Register SrcReg = MatchInfo;
+
+  B.buildCopy(DstReg, SrcReg);
+  MI.eraseFromParent();
+}
+
+/// Match something like this:
+/// %209:_(<16 x s32>) = G_INTRINSIC intrinsic(@llvm.aie2.concat.I512.I256),
+///         %95(<8 x s32>), %98(<8 x s32>)
+/// %252:_(<8 x s32>), %253:_(<8 x s32>) = G_UNMERGE_VALUES %209(<16 x s32>)
+///
+/// To convert to:
+/// 252:_(<8 x s32>) = COPY %95(<8 x s32>)
+/// 253:_(<8 x s32>) = COPY %98(<8 x s32>)
+bool llvm::matchUnmergeConcat(MachineInstr &MI, MachineRegisterInfo &MRI,
+                              const AIEBaseInstrInfo &TII,
+                              std::pair<MachineInstr *, unsigned> &MatchInfo) {
+
+  MachineInstr &SrcMI =
+      *MRI.getVRegDef(MI.getOperand(MI.getNumOperands() - 1).getReg());
+
+  std::optional<const AIEBaseInstrInfo::VConcatOpInfo> ConcatOp =
+      TII.getVConcatOpInfo(SrcMI);
+
+  if (!ConcatOp)
+    return false;
+
+  // We always have more operands for the intrinsic.
+  if (MI.getNumOperands() !=
+      SrcMI.getNumOperands() - ConcatOp->NumOfNonRegOperands)
+    return false;
+
+  MatchInfo = std::make_pair(&SrcMI, ConcatOp->FirstOperand);
+  return true;
+}
+
+void llvm::applyUnmergeConcat(MachineInstr &MI, MachineRegisterInfo &MRI,
+                              MachineIRBuilder &B,
+                              std::pair<MachineInstr *, unsigned> &MatchInfo) {
+  B.setInstrAndDebugLoc(MI);
+  auto [ConcatMI, Offset] = MatchInfo;
+
+  for (unsigned Op = 0; Op < MI.getNumOperands() - 1; Op++) {
+    Register DstReg = MI.getOperand(Op).getReg();
+    Register SrcReg = ConcatMI->getOperand(Op + Offset).getReg();
+    B.buildCopy(DstReg, SrcReg);
+  }
+
+  MI.eraseFromParent();
+}
+
+/// This function tracks chain of vector updates using .upd vector intrinsic.
+static bool trackVectorUpdateChain(MachineInstr &MI, MachineRegisterInfo &MRI,
+                                   const AIEBaseInstrInfo &TII,
+                                   std::map<unsigned, Register> &IndexRegMap) {
```

Comment:
Super-nit: We could actually just return a `std::map<unsigned, Register>`, the map would be empty if we do not manage to track any update.